File: client/src/modules/nexotype/schemas/omics/organism.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const OrganismSchema = z.object({
  id: z.number(),
  ncbi_taxonomy_id: z.number().int(),
  scientific_name: z.string().min(1).max(100),
  common_name: z.string().min(1).max(100),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateOrganismSchema = z.object({
  ncbi_taxonomy_id: z.number().int(),
  scientific_name: z.string().min(1).max(100),
  common_name: z.string().min(1).max(100),
});

export const UpdateOrganismSchema = z.object({
  ncbi_taxonomy_id: z.number().int().optional(),
  scientific_name: z.string().min(1).max(100).optional(),
  common_name: z.string().min(1).max(100).optional(),
});

// ==========================================
// Types
// ==========================================

export type Organism = z.infer<typeof OrganismSchema>;
export type CreateOrganism = z.infer<typeof CreateOrganismSchema>;
export type UpdateOrganism = z.infer<typeof UpdateOrganismSchema>;

// ==========================================
// Response Types
// ==========================================

export type OrganismResponse = {
  success: boolean;
  data?: Organism;
  error?: string;
};

export type OrganismsResponse = {
  success: boolean;
  data?: Organism[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/omics/organism_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class OrganismCreate(BaseModel):
    """Schema for creating a new organism"""
    ncbi_taxonomy_id: int
    scientific_name: str
    common_name: str


class OrganismUpdate(BaseModel):
    """Schema for updating an organism"""
    ncbi_taxonomy_id: int = None
    scientific_name: str = None
    common_name: str = None


# ==========================================
# Response Schemas
# ==========================================

class OrganismDetail(BaseModel):
    """Schema for organism details"""
    id: int
    ncbi_taxonomy_id: int
    scientific_name: str
    common_name: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class OrganismListResponse(BaseModel):
    """Response schema for listing organisms"""
    success: bool
    data: list[OrganismDetail] | None = None
    count: int = 0
    error: str | None = None


class OrganismResponse(BaseModel):
    """Response schema for single organism operations"""
    success: bool
    data: OrganismDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
